ref-exchange/src/stable_swap/mod.rs

```diff
@@ -0,0 +1,1092 @@
+        for (index, value) in self.token_decimals.iter().enumerate() {
+            let factor = 10_u128
+                .checked_pow((TARGET_DECIMAL - value) as u32)
+                .unwrap();
+            c_amounts[index] *= factor;
+            c_current_amounts[index] *= factor;
+        }
+
+        let invariant = self.get_invariant();
+        let trade_fee = Fees::new(self.total_fee, &fees);
+
+        let (burn_shares, _) = invariant
+            .compute_lp_amount_for_withdraw(
+                &c_amounts,
+                &c_current_amounts,
+                self.shares_total_supply,
+                &trade_fee,
+            )
+            .expect(ERR67_LPSHARE_CALC_ERR);
+
+        burn_shares
+    }
+
+    /// Remove liquidity from the pool by fixed tokens-out,
+    /// allows to remove liquidity of a subset of tokens, by providing 0 in `amounts`.
+    /// Fee will be charged according to diff between ideal token portions.
+    pub fn remove_liquidity_by_tokens(
+        &mut self,
+        sender_id: &AccountId,
+        amounts: Vec<Balance>,
+        max_burn_shares: Balance,
+        fees: &AdminFees,
+    ) -> Balance {
+        let n_coins = self.token_account_ids.len();
+        assert_eq!(amounts.len(), n_coins, "{}", ERR64_TOKENS_COUNT_ILLEGAL);
+        let prev_shares_amount = self.shares.get(&sender_id).expect(ERR13_LP_NOT_REGISTERED);
+
+        // make amounts into comparable-amounts
+        let mut c_amounts = amounts.clone();
+        let mut c_current_amounts = self.amounts.clone();
+        for (index, value) in self.token_decimals.iter().enumerate() {
+            let factor = 10_u128
+                .checked_pow((TARGET_DECIMAL - value) as u32)
+                .unwrap();
+            c_amounts[index] *= factor;
+            c_current_amounts[index] *= factor;
+        }
+
+        let invariant = self.get_invariant();
+        let trade_fee = Fees::new(self.total_fee, &fees);
+
+        let (burn_shares, fee_part) = invariant
+            .compute_lp_amount_for_withdraw(
+                &c_amounts,
+                &c_current_amounts,
+                self.shares_total_supply,
+                &trade_fee,
+            )
+            .expect(ERR67_LPSHARE_CALC_ERR);
+
+        assert!(
+            burn_shares <= prev_shares_amount,
+            "{}",
+            ERR34_INSUFFICIENT_LP_SHARES
+        );
+        assert!(burn_shares <= max_burn_shares, "{}", ERR68_SLIPPAGE);
+
+        for i in 0..n_coins {
+            self.amounts[i] = self.amounts[i].checked_sub(amounts[i]).unwrap();
+            assert!(
+                self.amounts[i] >= MIN_RESERVE
+                    .checked_mul(
+                        10_u128
+                            .checked_pow(self.token_decimals[i] as u32)
+                            .unwrap()
+                    )
+                    .unwrap(),
+                "{}",
+                ERR69_MIN_RESERVE
+            );
```

Comment:
This check should also be extracted to a separate method. E.g. assert_min_reserve()

Reply:
fixed.